### ISSUE TITLE
Update KH registers

### ISSUE THREAD

**Issue by canton7** [nathanmarlor/foxess_modbus#518]
We have the modbus spec now:

Fault codes, min/max doc, max charge/discharge current we don't map the holding registers for.

**Comment by lamixer** [comment]
Hello! I'm another KH owner and I have just (foolishly) updated my inverter firmware to manager 1.19 and the modbus integration stopped working. I updated to the latest beta (10b6) but this has not changed anything. I still have all entities blank in HA. Let me know if you need any help with probing registers or testing a fix.

**Comment by FozzieUK** [comment]
@lamixer try dropping back to 10.0b3

**Comment by lamixer** [comment]
Thanks FozzieUK, but I still get nothing on that version.

**Comment by FozzieUK** [comment]
Ok, I can see you're getting guidance over on FB - i'll leave it to Antony & Will 👍

**Comment by williamjeccles** [comment]
@canton7 / @FozzieUK I'm not sure about all the changes made to master recently but to restore my K series to working with home assistant I had to downgrade to 1.9.0 again, then manually apply the changes in this PR https://github.com/nathanmarlor/foxess_modbus/pull/512 downgrading to 10.0b3 didn't work for me.

**Comment by FozzieUK** [comment]
@williamjeccles Thanks Will, i'd been loosely following the KH changes and it looked like they were wrapped into the 10.0b3 link but it's obviously more complex than that 😊, i'll leave it to you guys 👍

**Comment by lamixer** [comment]
Hello ... I'm probably not adding anything new to the conversation now, but I after having my KH10.5 inverter updated to Manager 1.19 (etc) and losing all modbus entities, I have 'fixed' the problem with @williamjeccles help (Thank you!) by reverting to v1.90 of the integration and then manually modifying my installed version of inverter_profiles.py and entity_descriptions.py. 

I did not regain all of the entities, but most of them came back. The two I need to get my Sunsynk Flow card fully working again are EPS Power (FoxESS) and Load Power (FoxESS). 

I wonder if we should make a separate beta branch of the code to sort this out, or just carry on with the current beta sequence. I haven't read all of the comments but it seems something about the work to enable the new force charge capability has made it so these changes (from PR 512) which I merged in to 1.90 do not work in the 1.10b6 release.

**Comment by lamixer** [comment]
Oh, and besides the entites not working, 'battery charge total' is high by a factor of aprox 1000. Battery Discharge Total looks like it could be correct though.

Just to be thorough, the entties I don't have include:
(the below show 0)
Inverter Power
EPS Power
Battery Charge
Feed In (might be zero correctly for today so not sure)
Grid Consumption
Grid CT

(the below show Unavailable)
Battery Current
Battery Voltage
a bunch of BMS entries
All Period 1 and Period 2 entries

**Comment by canton7** [comment]
>  wonder if we should make a separate beta branch of the code to sort this out, or just carry on with the current beta sequence. I haven't read all of the comments but it seems something about the work to enable the new force charge capability has made it so these changes (from PR 512) which I merged in to 1.90 do not work in the 1.10b6 release.

It's mainly that I've been working 12 hour days and haven't had a chance to finish up! Project delivery is tomorrow, so hopefully things get back to normal soon.

**Comment by lamixer** [comment]
No worries Antony. Your efforts are much appreciated and you can only do what you can do!On 13 Feb 2024, at 22:03, Antony Male ***@***.***> wrote:﻿

wonder if we should make a separate beta branch of the code to sort this out, or just carry on with the current beta sequence. I haven't read all of the comments but it seems something about the work to enable the new force charge capability has made it so these changes (from PR 512) which I merged in to 1.90 do not work in the 1.10b6 release.

It's mainly that I've been working 12 hour days and haven't had a chance to finish up! Project delivery is tomorrow, so hopefully things get back to normal soon.

**Comment by michalzxc** [comment]
I dont know how, but for me it is not 41XXX but all needs 31XXX
```
./modpoll  -m tcp -p 502 -a 247 -t 3 -0 -1 -c 16 -r 31008  10.0.0.253 
modpoll 3.10 - FieldTalk(tm) Modbus(R) Master Simulator
Copyright (c) 2002-2021 proconX Pty Ltd
Visit https://www.modbusdriver.com for Modbus libraries and tools.

Protocol configuration: MODBUS/TCP, FC4
Slave configuration...: address = 247, start reference = 31008 (PDU), count = 16
Communication.........: 10.0.0.253, port 502, t/o 1.00 s, poll rate 1000 ms
Data type.............: 16-bit register, input register table

-- Polling slave...
[31008]: 0
[31009]: 5009
[31010]: 2494
[31011]: 9
[31012]: 0
[31013]: 4864
[31014]: 0
[31015]: 0
[31016]: 0
[31017]: 0
[31018]: 0
[31019]: 0
[31020]: 0
[31021]: 0
[31022]: 0
[31023]: 0
```

I have KH10.5 inverter 1.19 manager, 1.27 master

**Comment by canton7** [comment]
I know what needs to be done, don't worry. You don't need to do your own investigation.

**Comment by canton7** [comment]
Try v1.10.0b7, see if that improves anything.

I should have some time tomorrow to get this sorted.

**Comment by michalzxc** [comment]
🎉
Battery state is updating again! Thank you!!

**Comment by lamixer** [comment]
@williamjeccles Did you try the latest yet? I'm afraid to overwrite my hacked version that has most of the sensors working until I hear from you! Thanks.

**Comment by williamjeccles** [comment]
@canton7 / @lamixer v1.10.0b7 is working correctly for me

**Comment by canton7** [comment]
> @canton7 / @lamixer v1.10.0b7 is working correctly for me I can also confirm that force discharge is working for me on b7. Still can't see any grid or feedin values, I assume you're still working on those?

Those should be working. They moved the registers in the latest spec I've seen, but I fixed those in https://github.com/nathanmarlor/foxess_modbus/pull/535/files. I'll double check next time I'm at a computer.

**Comment by williamjeccles** [comment]
@canton7 you are correct, b7 has working ct values! Thanks!

**Comment by lamixer** [comment]
Hello ... thanks for the update. I installed v1.10.0b7 and it has improved the situation from v1.90 with manually-applied changes.

I'm sure this is a work in progress, but just reporting, from my above post noting missing entities I previously had, the lastest had changed the noted ones:

Just to be thorough, the entties I don't have include:
(the below show 0)
!! RESTORED -- Inverter Power
!! RESTORED -- EPS Power
Battery Charge
Feed In (might be zero correctly for today so not sure)
Grid Consumption
!! RESTORED -- Grid CT  but note, the +/- has swapped!

(the below show Unavailable)
Battery Current
Battery Voltage
a bunch of BMS entries
All Period 1 and Period 2 entries

Of course also still the "Charge Periods" setting card does not populate since the periods entities are missing.

Great step in the right direction, thank you!

**Comment by canton7** [comment]
> Battery Charge

You don't mention Battery Discharge or Inverter Battery Power. Are those OK? (all 3 entities are driven by the same register).

> Feed In (might be zero correctly for today so not sure)
> Grid Consumption
> !! RESTORED -- Grid CT but note, the +/- has swapped! 

Those 3 values are also driven from the same register. If Grid CT has swapped sign though, that might explain it. 

Can anyone else confirm whether Feed In / Grid Consumption are swapped? 

> Battery Current
> Battery Voltage
> a bunch of BMS entries 

FoxESS removed a bunch of BMS registers. I'll check which exactly next time I'm at a computer. 

> All Period 1 and Period 2 entries 

Those should still be there. Does HA sat that the integration had stopped providing them? Ant warnings in your log?

**Comment by canton7** [comment]
Right, I'm back and at a computer again!

> Actually, looking back at last night, I see sensor.foxess_battery_charge did register the overnight charging, so though I saw zero, it was a correct zero.

:+1:

> sensor.foxess_feed_in is showing zero but I am feeding in right now.
> sensor.foxess_grid_consumption is not working.

Ah yes, I have failed to map those, looking at it again. Will fix!

> Regarding binary_sensor.foxess_time_period_1_enable_charge_from_grid and similar, I get the message "This entity is no longer being provided by the foxess_modbus integration. If the entity is no longer in use, delete it in settings."

They're not actually listed in the latest spec I have for the KH. Can you check whether they're readable?

Go to Developer Tools -> Services, and read 3 input registers with a start address of 41001:

![image](https://github.com/nathanmarlor/foxess_modbus/assets/568104/00c4b59f-4ad1-46eb-a4b4-da2fb107ad25)

**Comment by lamixer** [comment]
No response from those registers. Any other locations we think it might be?

**Comment by canton7** [comment]
Trying reading them as holding registers, just to be sure?

That's where they are for the H1, and where they used to be for the KH. I can't see them anywhere in the latest spec I have, and so far that spec's been accurate.

**Comment by lamixer** [comment]
I tried input and holding- no response from either. 

I don’t mind going fishing. My setting for period 1 is force charge off, start time 23:30 end time 00:00 so what values should I look for and where do you suggest there is a contiguous set of registers it might be? Should I try input or holding when I poke it?

**Comment by canton7** [comment]
They'll be holding registers most likely. 23:30 is 5918. Make sure the charge period is actually enabled.

I've only ever seen them at 41001-41006.

**Comment by lamixer** [comment]
Can I see the register documentation that you have for KH and K series to give me a clue where to look?

I have seen one before but it sounds like you have an updated one.

**Comment by lamixer** [comment]
Thanks, that's a pretty nice document! Here's some probing:

31008 returns 0
31012 returns 0
31014-16 return 0
31054 returns 544 (this is aprox my house load right now)
31055-31059 no response
31060 - 31090 probing ten at a time produced no response (not sure best way to probe a big range)
We have 31055 to 31999 as possible locations to find charge period registers.
Ditto 32024 to 39999.
40006 returns 3
40007 returns 0
40008 returns 1
40009 returns 0
40010 returns 1
40011 rerturns 2
40012 returns 247
40013 returns 0
40014 - 40020 no response (read individually)
40100, 40200, 40300, 40400 no response
41001 - 41006 no response (read individually)
41012 returns 9000
41019 - 41030 no response (read individually)
42000 no response
43000 no response
44004 no response
44010 no repsonse
44100 no response
44200, 44300, 44400, 44500, 44600, 44700, 44800, 44900 no response
45005 no response.

Clearly, 40006 - 40013 could be interesting. For 40012 does 247 mean anything to you?

As noted, I only probed registers that were TBD or unmentioned in the document, so there are a few things showing up that may be not documented yet.

**Comment by lamixer** [comment]
9000 is my current (no pun intended) setting for Export Limited Power (range 0 - 30000) in W.
Maybe 247 refers to the modbus address.

**Comment by canton7** [comment]
> For 40012 does 247 mean anything to you?

Yep, Slave ID.

**Comment by lamixer** [comment]
Ok.. I reset the two charge periods to be both on and times in the 1am hour a few minutes apart. Now probing I get:

41014 = 1
41015 = 3
 then no response going up a few more registers.

I can't find any other changes resulting from this.

I don't have any strategy periods (or whatever those are called now) set, maybe will try and see if that pops up any new registers. 

Maybe @williamjeccles can enquire with FoxEss why they took away the working (but undocumented) charge period registers for KH in 1.19 manager!

**Comment by lamixer** [comment]
Setting something up in "mode scheduler" did not show up any new register data (that I could find anyway).

**Comment by canton7** [comment]
Yeah the mode scheduler isn't exposed over modbus currently (we know the "registers" it uses, there's a discussion on that, but they can't be read over modbus).

**Comment by lamixer** [comment]
@canton7 When you have a chance, can you please check sensor.foxess_battery_charge_total ? I think it's missing a k and I'm seeing watts instead of kWh. I'm not sure that explains it though, it seems it should be about the same as sensor.foxess_battery_discharge_total which is 2581.3 kWh whereas the charge total is showing 147,274,746.4 kWh.

**Comment by canton7** [comment]
> Would be great if there was an override the inverter would respond to (over modbus) that just says 'do this now' like charge at 10A or discharge to grid at 12A or stop doing either of those things and revert to stored settings. Then HA could easily trigger those things.

It does! That's what I implemented the new force charge / force discharge e work modes on top of.

**Comment by lamixer** [comment]
What I great idea I had (YOU HAD -- LOL)! Is it supposed to work on KH series?

When I put it in Backup mode it (within a second or two) reverts to Self-Use.
When I put it in either Force Charge or Force Discharge it starts discharging to the grid.

Whatever I can do to help it work on KH, please let me know. Maybe it has to do with the backward CT measurements and the other things you are working to fix for KH? (note, I have changed the polarity in my sunsynk flow card I use to monitor these things so I do know correctly which way the power is flowing to/from the grid. There's an invert grid sensor switch in that card.

**Comment by canton7** [comment]
Interesting. It works with the H1 and H3 currently. I wonder how the KH is different...

Can you enable debug logging, turn on force charge and force discharge, and send me the logs?

**Comment by canton7** [comment]
> When you have a chance, can you please check sensor.foxess_battery_charge_total 

Yep that was a bad definition thanks!

**Comment by canton7** [comment]
I'll release those fixes when we've figured out what's going on with the force charge / force discharge.

**Comment by lamixer** [comment]
I'm not sure how to change log levels. With my default settings, I get these results:

Self Use > Feed-In First
   May have worked (battery is at 95% so not sure)

Self Use > Backup
   Switched itself right back to Self Use

Self Use > Force Charge
  Stays set to FC but I see nothing change (my battery is 95% though)

Self Use > Force Discharge
  Works

No log entries in whatever default logging mode I'm in.

**Comment by canton7** [comment]
> Switched itself right back to Self Use

I think I've identified that as a separate issue. Leave it for a poll period (10 seconds by default), and it should start showing the new work mode.

**Comment by canton7** [comment]
Use this button to enable debug logging:

![image](https://github.com/nathanmarlor/foxess_modbus/assets/568104/d4083afc-51c9-44fe-9a70-51894872a1d3)

**Comment by lamixer** [comment]
Thank you... Again then with dub logging on:

Self Use > Feed-In First
May have worked (battery is at 95% so not sure)
No log entries generated.

Self Use > Backup
Switched itself right back to Self Use
No log entries generated.

Self Use > Force Charge
Begins to force discharge (battery was at 92%, I raised my Max SoC to 96% then it started discharging)
No log entries generated.

Self Use > Force Discharge
Begins to force discharge
No log entries generated.

However, I then downloaded the full log file from HA and there are entries. Attached!

[lamxer-21Feb2024.log](https://github.com/nathanmarlor/foxess_modbus/files/14358431/lamxer-21Feb2024.log)

**Comment by lamixer** [comment]
I'm seeing mention of 12000 watts there. My inverter will do 10500 as per spec, though I have some limits set at 9000 I believe.

**Comment by lamixer** [comment]
But when the force discharge was working there, it was respecting some setting I have somewhere of about 5kW.

**Comment by canton7** [comment]
That all seems to be fine. It's writing 53536 (-12000) to trigger a force charge. If the inverter actually starts to export at that point, I can only think that's a problem with the inverter. That's fun, more things to work around.

> I'm seeing mention of 12000 watts there. My inverter will do 10500 as per spec, though I have some limits set at 9000 I believe. Is it trying to do 12kW and failing?

It shouldn't -- that's not how the other inverters work. (I can't actually tell the inverter capacity for the KH, so I have to default to the largest capacity).

**Comment by lamixer** [comment]
Ok... 

The force charge is interesting, because it doens't start if the battery is already at the Max SoC. If I raise that setting then the inverter says, 'ok, I'll start charging then' except is actually DISCHARGES!

If you push a new beta with the other fixes maybe @williamjeccles and I can hop on that and double check things. Thanks!

**Comment by canton7** [comment]
> If I raise that setting then the inverter says, 'ok, I'll start charging then' except is actually DISCHARGES!

Yes. I can see that.

I only tell the inverter to force charge if the soc is <= the max soc (as force charge doesn't respect max soc).

And yes, as I said, when I tell the inverter to charge, you claim it's actually discharging.

**Comment by canton7** [comment]
Can you show me the graph for the rpower sensor when you turn on Force Charge and then Force Discharge?

**Comment by lamixer** [comment]
they are brief since I don't really want to discharge my battery right now .. but you can see them there.
![rpower-graph1](https://github.com/nathanmarlor/foxess_modbus/assets/4958776/87e8c7b8-85de-4be1-9bc4-f607377ce426)

**Comment by canton7** [comment]
Thanks. Can you try setting Force Charge Power to something low, like 0.5kW, and trying force charge again.

**Comment by lamixer** [comment]
Sure, but not sure where that setting is. Do you mean number.foxess_max_charge_current ?

**Comment by lamixer** [comment]
Oh... finding it now!

**Comment by canton7** [comment]
![image](https://github.com/nathanmarlor/foxess_modbus/assets/568104/8acac91e-f7d5-4eec-bd39-d9e52a124edb)

**Comment by lamixer** [comment]
Yes, sorry. I didn't have that in my little 'control panel' I made. It was set to 12kW. Let me try now.

**Comment by williamjeccles** [comment]
![image](https://github.com/nathanmarlor/foxess_modbus/assets/6324545/6c2341f4-2c8d-4bb2-a04e-b9244e2a14b2)
Force charge/discharge is working for me

**Comment by williamjeccles** [comment]
I have the charge from grid setting set to 10amps which is overriding the normal max charge amps which for me is fine but maybe a setting you can expose @canton7 ? 

![image](https://github.com/nathanmarlor/foxess_modbus/assets/6324545/ce3921b7-31ec-4371-9aed-1f619f1fa989)

**Comment by lamixer** [comment]
I set it to 1kW and it worked perfectly (I think) ... certainly works. I suggest you make the default value the LOWEST anyone could have (smallest inverter capacity) then leave it up to the user to raise it -- since it fails is a bad way when it is set too high. Wow... this is great!

**Comment by canton7** [comment]
@williamjeccles Thanks! What's your inverter capacity?

> since it fails in a bad way when it is set too high

It seems to only fail on the KH. Setting it to some arbitrary low value is going to generate lots of questions in here, as people won't understand why it's charging/discharging at some arbitrary low value. I'll see what else I can do....

> I have the charge from grid setting set to 10amps which is overriding the normal max charge amps which for me is fine but maybe a setting you can expose @canton7 ?

I can't see a register for that parameter anywhere in the spec. 

You're saying that the inverter is respecting GridChargeCurrentLimited? When would we not want to respect that value?

**Comment by lamixer** [comment]
Great, thanks! Initial results for v1.10.0b9:

The following entities are missing that previously existed:

```
sensor.foxess_bat_current
```

missing but not sure ever existed (but HA thinks so)
```
sensor.foxess_batvolt
sensor.foxess_inverter_fault_code
sensor.foxess_bms_cycle_count
sensor.foxess_bms_kwh_remaining
```

Earlier today I deleted some other formerly-known to HA entities in BMS category, so the above list is incomplete from v1.9 and my earlier inverter manager version. Let me know if any of those are not expected to return and I'll delete them. I would very much like bat_current to return though.

On the plus side, `sensor.foxess_battery_charge_total` looks correct now!

Force discharge works. Force charge works. Backup mode still pops back to self-use (even if it was set when on a mode other than self-use). Feed-in First setting will stick, but at this time I can't determine if it works.

The 'polarity' (or "sign") of the Grid CT is still swapped from how it was. I'm not sure if that's a problem or not. When I go into 'Force Charge' and my battery drawing from the grid for sure, the entity `sensor.foxess_grid_ct` is POSITIVE.

**Comment by canton7** [comment]
Yep, those have been removed by FoxESS, [see above](https://github.com/nathanmarlor/foxess_modbus/issues/518#issuecomment-1952425731).

(The fault code is still there, but the meanings of the different bits has changed. I'll need to go and re-map them at some point, at which point that sensor will reappear).

You still have Inverter Battery Current, which is the current leaving the inverter for the batteries ([see here](https://github.com/TonyM1958/HA-FoxESS-Modbus/wiki/How-does-it-work%3F)).

> The polarity (or "sign") of the Grid CT is still swapped from how it was. I'm not sure if that's a problem or not. When I go into 'Force Charge' and my battery is drawing from the grid for sure, the entity sensor.foxess_grid_ct is POSITIVE.

How odd. I assume "Feed In" and "Grid Consumption" are the wrong way around as well?

**Comment by lamixer** [comment]
Yes, since the changes Feed In and Grid Consumption have been stuck at zero, except actually just since today's update they are showing again. 

Also just confirmed, when I Force-Charge, Feed-In goes up. When I Force-Discharge Grid Consumption goes up.

**Comment by williamjeccles** [comment]
> @williamjeccles Thanks! What's your inverter capacity?
> 
> > I suggest you make the default value the LOWEST anyone could have (smallest inverter capacity) then leave it up to the user to raise it -- since it fails in a bad way when it is set too high
> 
> It seems to only fail on the KH. Setting it to some arbitrary low value is going to generate lots of questions in here, as people won't understand why it's charging/discharging at some arbitrary low value. I'll see what else I can do....
> 
> > I have the charge from grid setting set to 10amps which is overriding the normal max charge amps which for me is fine but maybe a setting you can expose @canton7 ?
> 
> I can't see a register for that parameter anywhere in the spec.
> 
> You're saying that the inverter is respecting GridChargeCurrentLimited? When would we not want to respect that value?

Hi @canton7 yes, my KH10.5 is respecting the value for GridChargeCurrentLimited for grid charging which is expected. I'm using a 10.5kW inverter and have no issues with the default of 10kW but then again I have a large inverter. KH start at 7kW so maybe default it to 7kW for the KH series.

**Comment by canton7** [comment]
> Hi @canton7 yes, my KH10.5 is respecting the value for GridChargeCurrentLimited for grid charging which is expected. I'm using a 10.5kW inverter and have no issues with the default of 10kW but then again I have a large inverter. KH start at 7kW so maybe default it to 7kW for the KH series.

Ta! I ended up parsing the inverter capacity out of the model name -- hopefully that will be good enough!

**Comment by williamjeccles** [comment]
@canton7 b10 works for me, 
Grid consumption - Working
Feedin - Working
CT - appears to be correctly reporting now

i'm not sure EPS power is correct but will confirm when i'm home

**Comment by lamixer** [comment]
Looks good here as well. 

Could you please tell me how to use an automation to set the work mode? I had it working to set the Min SOC (on grid) to pull from grid when the car charger started up. I used "Number: Set". I don't know which to choose there to set the mode.

I do now (and have since the inverter manager update) received errors if using the integration to set the Min SoC (on grid):

`[139819738488768] Error writing registers. Start: 41011; values: [94]; slave: 247 from tcp://192.168.1.18:502: Exception Response(134, 6, IllegalValue)`

**Comment by canton7** [comment]
> Could you please tell me how to use an automation to set the work mode?

The same as any other number entity.

![image](https://github.com/nathanmarlor/foxess_modbus/assets/568104/34e64e97-8601-4553-8ac1-af40e86ba918)

> I do now (and have since the inverter manager update) received errors if using the integration to set the Min SoC (on grid)

Interesting. @williamjeccles can you repro?

**Comment by lamixer** [comment]
I want to set the Work Mode so I can use your new efforts in an automation. If I choose to set a number I cannot set the work mode because it is not a number. I thought Text: Set Value might work, but there are no entities that way.

**Comment by lamixer** [comment]
Right! Maybe you can create a new service to make that easier for others to figure out, like you already have "Update Charge Period" etc. Will also prevent typos using Select and thus having to type in the name of the mode you want correctly. I think I have done it, will find out next time Octopus wants to charge my car! Thank you!

**Comment by lamixer** [comment]
to clarify, the error happened when I changed "number.foxess_max_soc" but I think this may be a transient failure because earlier today I used it and it worked. (scratching head)

**Comment by lamixer** [comment]
Yes.. it's working!

**Comment by lamixer** [comment]
Oh wow... I poked around before and found nothing like that! I think it's because I was editing an existing action.

But now I did. To help anyone else, when you are in the GUI to create an automation, when you click to Add Action you can click on Device, then start typing Fox to find the FoxESS modbus device, then actions are pre-populated, choose "Change Work Mode (FoxESS) option" then you'll see the work modes you can choose like the screenshot above.

**Comment by lamixer** [comment]
I can confirm that using the integration to change work mode to 'feed in first' results in the FoxEss cloud interface also showing that mode also. Looks to be working.

**Comment by lamixer** [comment]
Here is verbose log of changing from self-use to backup mode (failure).
[lamixer-22Feb2024-BackupModeChange.log](https://github.com/nathanmarlor/foxess_modbus/files/14371869/lamixer-22Feb2024-BackupModeChange.log)

**Comment by canton7** [comment]
That log only has attempts to write min soc (which are failing because your inverter isn't respondong.

**Comment by lamixer** [comment]
Sorry, I assumed when I clicked debug logging on then off and it downloaded the log to my computer it was only the most recent run ... but it had older stuff in it.

Attached is a log from about a minute and a half where I twice attempted to go to backup mode (and it reverted to self-use). Just in case it helps figure out that issue. If Backup mode worked for me I could turn off the charge periods and just use the modes triggered by time and conditions to control everything I need to. Even without backup mode working, this is reallly great to have now. Thanks again!

[home-assistant_foxess_modbus_2024-02-22T12-42-34.134Z.log](https://github.com/nathanmarlor/foxess_modbus/files/14374018/home-assistant_foxess_modbus_2024-02-22T12-42-34.134Z.log)

**Comment by canton7** [comment]
Can you try writing the value 2 to register 41000 with modpoll, then reading back register 41000 and seeing what it contains?

It does look like it's getting written fine, but presumably the next read back is giving back a different number.

**Comment by lamixer** [comment]
I used the HA gui Developer Tools service to write and read it. No error reported on the write. Read back value of 0.

**Comment by lamixer** [comment]
Not just from the cloud, I should say, even from the front panel control there are only the two modes.

**Comment by williamjeccles** [comment]
> > Could you please tell me how to use an automation to set the work mode?
> 
> The same as any other number entity.
> 
> ![image](https://private-user-images.githubusercontent.com/568104/306711451-34e64e97-8601-4553-8ac1-af40e86ba918.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg2MTM5MTQsIm5iZiI6MTcwODYxMzYxNCwicGF0aCI6Ii81NjgxMDQvMzA2NzExNDUxLTM0ZTY0ZTk3LTg2MDEtNDU1My04YWMxLWFmNDBlODZiYTkxOC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjIyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIyMlQxNDUzMzRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02MjI5OTVlNTFmMmZiMGRiYmVlYjQxN2NlNWJmMzQzMjhjNDJlZjUxNmJjNDgyM2Y1N2FhNTYwOWE1MTg0MGY5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.RSqF1NRJ4g6S2CTtAJbsIpEp_k6TfAhY_U5CyGCgDUg)
> 
> > I do now (and have since the inverter manager update) received errors if using the integration to set the Min SoC (on grid)
> 
> Interesting. @williamjeccles can you repro?

![image](https://github.com/nathanmarlor/foxess_modbus/assets/6324545/44b5d6c1-a018-4292-a26c-0f5f15a1f215)
 mine is working fine

**Comment by williamjeccles** [comment]
Remember there is new KH firmware coming soon

**Comment by canton7** [comment]
@williamjeccles Do you know when it's due to arrive? I'm wondering whether to hold off on the 1.10 release, or get in a workaround for the lack of Back-Up work mode.

**Comment by williamjeccles** [comment]
@canton7 I was told March about 2 weeks ago.

**Comment by canton7** [comment]
@williamjeccles It's mapped to the EPS Power register. Do you have a better register to map it to?

![image](https://github.com/nathanmarlor/foxess_modbus/assets/568104/810d7f1e-a937-415f-9822-f0d3f4a17941)

**Comment by lamixer** [comment]
sensor.foxess_eps_rpower is what I'm using ... it seems correct. Note the r there.

**Comment by williamjeccles** [comment]
Odd, my Master is 0 watts but I do see what I expect on my slave for the eps data. Going to check my wiring of the EPS side today.

**Comment by williamjeccles** [comment]
tested and it's working fine, was just my wiring.

**Comment by canton7** [comment]
Good to hear!

Leaving open: fault codes are still tracked by this issue.

**Comment by williamjeccles** [comment]
oh oops, no worries.

**Comment by cperricos** [comment]
Hello, I just had my solar PV installed with a Fox KH10 and got the modbus integration working to you amazing folk. My master version firmware is 1.18, and my manager version is 1.13. I have looked through this forum but just wanted to double check. is it safe/worth upgrading to latest firmware or do the problems with the news firmware still exist?

Also I have tried changing the polling intervals of the data, but it doesnt change very quickly (I need to create a load shedding solution as I am n a single phase supply so need a quick sensor on the house load). Does anyone know how quickly the modbus data is updated by the KH10?

thanks
